File: Estruturas_de_Dados/abstratos/Fila/fila.py
class Queue_Node:

    def __init__(self, data):
        self.data = data
        self.next = None

    def __repr__(self):
        return str(self.data)

class Queue():
    def __init__(self, maximum_size = 255, head:Queue_Node = None) -> None:
        
        self.head = head
        self.tail = head
        self.maximum_size = maximum_size
        self.number_elements = 0
                
    def show_queue(self):
        
        if self.number_elements > 0:
            print("\nFila: [", end="")
            
            temp_node = self.head

            while temp_node:

                print(temp_node, end = ", ")
                temp_node = temp_node.next

            print("\b\b]\n")
        
        else:
            print("A fila está vazia")
        
        return
            
    def add_queue(self, value):
        
        # A Fila é bastante parecida com a lista encadeada na hora de 
        # adicionar itens, fazendo uso das mesmas variaveis head e tail,
        # porém, adicionando no tail ao invés do head.

        if self.maximum_size > self.number_elements:        
            
            new_node = Queue_Node(value)

            if self.tail:
                self.tail.next = new_node
                self.tail = new_node
            else:
                self.head = new_node
                self.tail = new_node
            self.number_elements += 1

            return
            
        else:
            print(f"Tentou colocar node de valor {value} mas a fila está cheia")
            return
        
    def pop_queue(self):
        
        # Head aqui, é reservado pros valores que tão no ínicio da fila 
        # e que tão prestes a sair.

        if self.number_elements > 0:

            removed_node = self.head
            self.head = self.head.next
            self.number_elements -= 1

            print(f"Node de valor {removed_node.data} retirado da fila")
        
        else:
            print("A fila está vázia")

if __name__ == "__main__":

    fila = Queue(4)

    fila.pop_queue()

    fila.add_queue(25)
    fila.add_queue(58)
    fila.add_queue(3)

    fila.show_queue()

    fila.add_queue(86)
    fila.add_queue(44)

    fila.show_queue()

    fila.pop_queue()
    fila.pop_queue()

    fila.add_queue(883)

    fila.show_queue()

    fila.pop_queue()
    fila.pop_queue()

    fila.show_queue()

    # <- [25, 58, 3, 86] <-

    # <- [58, 3, 86] <-

    # <- [58, 3, 86, 883] <-

    # <- [3, 86, 883] <-

File: Estruturas_de_Dados/abstratos/Fila/Fila.java
package Complexo.Fila;

public class Fila {
    
    public static void main(String[] args) {

        Queue fila = new Queue(4);

        fila.popQueue();

        fila.showQueue();

        fila.addQueue("Justin");
        fila.addQueue("Marcos");
        fila.addQueue("Narcisa");

        fila.showQueue();

        fila.popQueue();
        fila.popQueue();

        fila.addQueue("Roberto");
        fila.addQueue("Rodrigo");

        fila.showQueue();

        fila.popQueue();

        System.err.println(fila.popQueue());

        fila.showQueue();
}

}

class Queue {

    QueueNode head;
    QueueNode tail;
    int maximumSize;
    int numberElements = 0;

    Queue(int maximumSize, QueueNode head) {

        this.maximumSize = maximumSize;
        this.head = head;
        this.tail = head;

    };

    Queue(int maximumSize) {

        this(maximumSize, null);

    }

    Queue() {

        this(255, null);
    }

    void showQueue() {

        if (this.numberElements > 0) {
            System.out.print("[");

            QueueNode tempNode = this.head;

            while (tempNode != null) {

                System.err.print(tempNode.data + ", ");
                tempNode = tempNode.next;

            }

            System.out.println("\b\b]");

    } else {
        System.err.println("A fila está vazia");
        
    }

}

    void addQueue(String data) {

        if (this.maximumSize > this.numberElements) {

            QueueNode newNode = new QueueNode(data);

            if (this.tail != null) {
                this.tail.next = newNode;
                this.tail = newNode;
            } else {
                this.head = newNode;
                this.tail = newNode;
            }
            this.numberElements++;
        }
    }

    String popQueue() {
        if (this.numberElements > 0) {

            QueueNode removedNode = this.head;            

            this.head = this.head.next;
            this.numberElements--;
            
            return removedNode.data;
        } else {
            System.out.println("\nA fila está vazia");
            return null;
        }
    }
    
}

class QueueNode {

    String data;
    QueueNode next;

    public QueueNode(String data) {

        this.data = data;
        this.next = null;

    }

}